### PR TITLE
Pre-commit config, linting, and type hints

### PR DESCRIPTION
**Pull request by sondrelg**
Hi @jerry-git 🙂 

I'm going to sit down and have a go at writing the implementation suggested in https://github.com/jerry-git/pytest-split/issues/11 to see if it works. 

I noticed that there was a pre-commit config in the project but that it maybe wasn't fully functional, so I've updated it (let me know if you prefer to remove some of the hooks I've added), run it on the project, and added type hints to the project. Hope that's something you want.

Opening this as an initial PR so it doesn't affect the diff for the next PR(s) 🙂 

If you want, I could also add a [linting workflow](https://github.com/snok/pep585-upgrade/blob/master/.github/workflows/test.yml#L6) and/or test workflow to the project to run on pull-requests

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -1,2 +1,4 @@
 [flake8]
-max-line-length = 88
\ No newline at end of file
+max-line-length = 88
+exclude =
+    tests/*
```
Comment (jerry-git) [inline setup.cfg:4]: what's the motivation behind not linting tests?

File: setup.py [hunk 2]
```diff
@@ -25,14 +22,18 @@
...
-        "Development Status :: 1 - Planning",
+        "Development Status :: 4 - Beta"
         "Intended Audience :: Developers",
         "Operating System :: OS Independent",
         "Programming Language :: Python :: 3",
         "Programming Language :: Python :: 3.6",
         "Programming Language :: Python :: 3.7",
         "Programming Language :: Python :: 3.8",
+        "Programming Language :: Python :: 3.9",
+        "Programming Language :: Python :: 3.10",
```
Comment (jerry-git) [inline setup.py:33]: Let's add these only when they are tested in the CI

File: src/pytest_split/plugin.py [hunk 3]
```diff
@@ -1,14 +1,23 @@
 import json
 import os
 from collections import defaultdict, OrderedDict
+from typing import TYPE_CHECKING, Tuple
 
 from _pytest.config import create_terminal_writer
+from _pytest.main import Session
+
+if TYPE_CHECKING:
```
Comment (jerry-git) [inline src/pytest_split/plugin.py:9]: If we introduce type hints, let's introduce type checking as well. For example, `mypy` pre-commit hook would be nice 🙂

**Comment by jerry-git** [comment]
> If you want, I could also add a linting workflow and/or test workflow to the project to run on pull-requests

Sure that would be nice. And it would be actually great if GHA would be used for all the CI related activities.

File: src/pytest_split/plugin.py (hunk 3 above)
Comment (jerry-git) [reply src/pytest_split/plugin.py:9]: or is `flake8-type-checking` running mypy under the hood?

Comment (sondrelg) [reply src/pytest_split/plugin.py:9]: No it doesn't - I agree, I'll add mypy!

File: setup.py (hunk 2 above)
Comment (sondrelg) [reply setup.py:33]: I'll add the workflows to this PR :)

File: setup.cfg (hunk 1 above)
Comment (sondrelg) [reply setup.cfg:4]: We should lint tests, I just had so many errors I originally decided to opt-out, but I can fix them instead 🙂

Comment (sondrelg) [reply setup.cfg]: Added back in now 👍

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Install Pytest ${{ matrix.pytest-version }} and pytest-cov
+        run: pip install pytest==${{ matrix.pytest-version }} pytest-cov
+      - name: Install package
+        run: pip install -e .
+      - name: Run tests
+        run: |
+          pytest --cov=pytest_diff --verbose --assert=plain
```
Comment (jerry-git) [inline .github/workflows/test.yml:48]: Aren't we now running the tests with both Travis and GHA?

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -1,10 +1,37 @@
...
+            'flake8-comprehensions',
+            'flake8-deprecated',
+            'flake8-print',
+            'flake8-type-checking',
+        ]
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: 'v0.812'
+    hooks:
+      - id: mypy
+        additional_dependencies: [django, djangorestframework, inflection, openapi-spec-validator, prance, pyYAML, django-stubs, djangorestframework-stubs, drf_yasg, drf-spectacular]
```
Comment (jerry-git) [inline .pre-commit-config.yaml:37]: are you sure django&friends are needed here 😅

File: .github/workflows/test.yml (hunk 4 above)
Comment (sondrelg) [reply .github/workflows/test.yml]: I've never used Travis. How is that configured? Don't see any traces of it in the repo or elsewhere 😛

File: .pre-commit-config.yaml (hunk 5 above)
Comment (sondrelg) [reply .pre-commit-config.yaml]: I'm sure they're not! 😄

File: .github/workflows/test.yml (hunk 4 above)
Comment (jerry-git) [reply .github/workflows/test.yml]: See https://travis-ci.org/github/jerry-git/pytest-split/branches and https://github.com/jerry-git/pytest-split/blob/master/.travis.yml

**Comment by sondrelg** [comment]
> > If you want, I could also add a linting workflow and/or test workflow to the project to run on pull-requests
> 
> Sure that would be nice. And it would be actually great if GHA would be used for all the CI related activities.

Workflow now seems to work (see [here](https://github.com/sondrelg/pytest-split/actions/runs/892070450)) 🎉 

I removed tests from the flake8-ignore list, added mypy, and added configs for the flake8-packages used 🙂

**Review by jerry-git — APPROVED** [review]
Great stuff 💪 

Thank you!

**Pull request merged**